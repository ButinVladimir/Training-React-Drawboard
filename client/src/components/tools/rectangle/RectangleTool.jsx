import React from 'react';
import BaseTool from '../BaseTool';
import RectangleToolOptions from './RectangleToolOptionsContainer';

const RECTANGLE_TOOL_NAME = 'Rectangle';

class RectangleTool extends BaseTool {
  constructor() {
    super();

    this.onSelectWidth = this.onSelectWidth.bind(this);
    this.onSelectColor = this.onSelectColor.bind(this);

    this.width = 1;
    this.color = '#000000';
  }

  get name() {
    return RECTANGLE_TOOL_NAME;
  }

  onSelectWidth(width) {
    this.width = width;
  }

  onSelectColor(color) {
    this.color = color;
  }

  get optionsElement() {
    return (
      <RectangleToolOptions
        width={this.width}
        color={this.color}
        onSelectWidth={this.onSelectWidth}
        onSelectColor={this.onSelectColor}
      />
    );
  }
}

export default RectangleTool;
