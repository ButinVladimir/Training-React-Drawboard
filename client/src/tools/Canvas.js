export const DEFAULT_DELTA_X = 0.0;
export const DEFAULT_DELTA_Y = 0.0;
export const DEFAULT_ZOOM = 1.0;
export const DEFAULT_ROTATION = 0.0;
export const DEFAULT_WIDTH = 500;
export const DEFAULT_HEIGHT = 500;
export const DEFAULT_CANVAS_COLOR = '#fff';
export const BORDER_COLOR = '#999';
export const BORDER_WIDTH = 3;

class Canvas {
  constructor() {
    this.deltaX = DEFAULT_DELTA_X;
    this.deltaY = DEFAULT_DELTA_Y;
    this.zoom = DEFAULT_ZOOM;
    this.rotation = DEFAULT_ROTATION;

    this.containerPositionX = 0;
    this.containerPositionY = 0;

    this.screenCanvasValue = null;
    this.screenContext = null;

    this.htmlCanvasElement = document.createElement('canvas');
    this.htmlCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.htmlCanvasElement.setAttribute('height', DEFAULT_HEIGHT);

    this.htmlCanvasContext = this.htmlCanvasElement.getContext('2d');
    this.htmlCanvasContext.fillStyle = DEFAULT_CANVAS_COLOR;
    this.htmlCanvasContext.fillRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.bufferCanvasElement = document.createElement('canvas');
    this.bufferCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.bufferCanvasElement.setAttribute('height', DEFAULT_HEIGHT);
    this.bufferCanvasContext = this.bufferCanvasElement.getContext('2d');
  }

  get screenCanvas() {
    return this.screenCanvasValue;
  }

  set screenCanvas(value) {
    this.screenCanvasValue = value;
    this.screenContext = value.getContext('2d');
  }

  outputCompletely() {
    this.screenContext.resetTransform();
    this.screenContext.clearRect(0, 0, this.screenCanvas.width, this.screenCanvas.height);

    this.show();
  }

  changeDelta(deltaX, deltaY) {
    this.hide();
    this.deltaX = deltaX;
    this.deltaY = deltaY;
    this.show();
  }

  changeZoom(zoom) {
    this.hide();
    this.zoom = zoom;
    this.show();
  }

  changeRotation(rotation) {
    this.hide();
    this.rotation = rotation;
    this.show();
  }

  applyTransformations() {
    this.screenContext.resetTransform();
    this.screenContext.translate(this.deltaX, this.deltaY);
    this.screenContext.rotate(this.rotation);
    this.screenContext.scale(this.zoom, this.zoom);
  }

  hide() {
    this.applyTransformations();

    this.screenContext.clearRect(
      -2 * BORDER_WIDTH, -2 * BORDER_WIDTH,
      DEFAULT_WIDTH + 4 * BORDER_WIDTH,
      DEFAULT_HEIGHT + 4 * BORDER_WIDTH,
    );
  }

  show() {
    this.applyTransformations();

    this.screenContext.strokeStyle = BORDER_COLOR;
    this.screenContext.lineWidth = BORDER_WIDTH;
    this.screenContext.strokeRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.screenContext.drawImage(this.htmlCanvasElement, 0, 0);
  }

  clearBuffer() {
    this.bufferCanvasContext.resetTransform();
    this.bufferCanvasContext.clearRect(
      0,
      0,
      this.bufferCanvasElement.width,
      this.bufferCanvasElement.height,
    );
  }

  updateAndShowBuffer(updateCb, ...args) {
    this.clearBuffer();

    this.bufferCanvasContext.scale(1 / this.zoom, 1 / this.zoom);
    this.bufferCanvasContext.rotate(-this.rotation);
    this.bufferCanvasContext.translate(-this.deltaX, -this.deltaY);
    this.bufferCanvasContext.translate(-this.containerPositionX, -this.containerPositionY);

    updateCb(this.bufferCanvasContext, ...args);

    this.show();
    this.screenContext.drawImage(this.bufferCanvasElement, 0, 0);
  }

  appendBuffer() {
    this.htmlCanvasContext.resetTransform();
    this.htmlCanvasContext.drawImage(this.bufferCanvasElement, 0, 0);
    this.clearBuffer();

    this.show();
  }
}

export default Canvas;
