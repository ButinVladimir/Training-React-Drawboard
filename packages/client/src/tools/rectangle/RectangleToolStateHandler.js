import BaseToolStateHandler from '../BaseToolStateHandler';

class RectangleToolStateHandler extends BaseToolStateHandler {
  onMouseMove(toolState) {
    this.canvas.updateAndShowBuffer(this, toolState);
  }

  onMouseUp(toolState) { // eslint-disable-line class-methods-use-this
    this.canvas.appendBuffer();
    toolState.fixRelativeCoords(this.canvas.containerPositionX, this.canvas.containerPositionY);

    return toolState.serialize();
  }

  applyState(context, viewState, toolState) { // eslint-disable-line class-methods-use-this
    if (toolState.addFill) {
      context.fillStyle = toolState.fillColor; // eslint-disable-line no-param-reassign
      context.fillRect(
        toolState.anchorX,
        toolState.anchorY,
        toolState.clientX - toolState.anchorX,
        toolState.clientY - toolState.anchorY,
      );
    }

    if (toolState.addBorder) {
      const lineWidth = toolState.borderWidth * viewState.zoom;
      context.lineWidth = lineWidth; // eslint-disable-line no-param-reassign
      context.strokeStyle = toolState.borderColor; // eslint-disable-line no-param-reassign
      context.strokeRect(
        toolState.anchorX,
        toolState.anchorY,
        toolState.clientX - toolState.anchorX,
        toolState.clientY - toolState.anchorY,
      );
    }
  }
}

export default RectangleToolStateHandler;
