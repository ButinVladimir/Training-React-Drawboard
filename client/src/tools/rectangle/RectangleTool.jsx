import React from 'react';
import BaseTool from '../BaseTool';
import RectangleToolState from './RectangleToolState';
import RectangleToolStateHandler from './RectangleToolStateHandler';
import BaseFigureToolOptions from '../../components/tools/figure/BaseFigureToolOptionsContainer';

const RECTANGLE_TOOL_NAME = 'Rectangle';

class RectangleTool extends BaseTool {
  constructor(canvas) {
    super(canvas);

    this.toolState = new RectangleToolState();
    this.stateHandler = new RectangleToolStateHandler(canvas);
  }

  static get name() {
    return RECTANGLE_TOOL_NAME;
  }

  get optionsElement() {
    return (
      <BaseFigureToolOptions
        key={RECTANGLE_TOOL_NAME}
        toolState={this.toolState}
      />
    );
  }

  unserializeState(stateObj) { // eslint-disable-line class-methods-use-this
    return RectangleToolState.unserialize(stateObj);
  }
}

export default RectangleTool;
