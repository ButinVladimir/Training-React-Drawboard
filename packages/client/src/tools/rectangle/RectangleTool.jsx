import React from 'react';
import BaseTool from '../BaseTool';
import RectangleToolState from './RectangleToolState';
import RectangleToolStateHandler from './RectangleToolStateHandler';
import BaseFigureToolOptions from '../../components/tools/figure/BaseFigureToolOptionsContainer';

const RECTANGLE_TOOL_NAME = 'Rectangle';

class RectangleTool extends BaseTool {
  constructor(canvas) {
    super(canvas, new RectangleToolState(), new RectangleToolStateHandler(canvas));
  }

  static get name() {
    return RECTANGLE_TOOL_NAME;
  }

  get optionsElement() {
    return (
      <BaseFigureToolOptions
        key={RECTANGLE_TOOL_NAME}
        toolState={this.toolState}
      />
    );
  }

  deserializeState(stateObj) { // eslint-disable-line class-methods-use-this
    return RectangleToolState.deserialize(stateObj);
  }
}

export default RectangleTool;
