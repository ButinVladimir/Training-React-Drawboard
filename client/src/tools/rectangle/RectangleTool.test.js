import { shallow } from 'enzyme';
import RectangleTool from './RectangleTool';
import RectangleToolState from './RectangleToolState';
import RectangleToolStateHandler from './RectangleToolStateHandler';
import Canvas from '../Canvas';

describe('RectangleTool', () => {
  it('is instantiated properly', () => {
    const canvas = new Canvas();
    const rectangleTool = new RectangleTool(canvas);

    expect(rectangleTool.canvas).toBeInstanceOf(Canvas);
    expect(rectangleTool.toolState).toBeInstanceOf(RectangleToolState);
    expect(rectangleTool.stateHandler).toBeInstanceOf(RectangleToolStateHandler);
  });

  it('returns name', () => {
    expect(RectangleTool.name).toBe('Rectangle');
  });

  it('returns options element', () => {
    const rectangleTool = new RectangleTool(null);
    const options = shallow(rectangleTool.optionsElement);

    expect(options).toMatchSnapshot();
  });

  it('returns deserialized state', () => {
    const serializedObject = {
      anchorX: 1,
      anchorY: 2,
      clientX: 3,
      clientY: 4,
      addBorder: true,
      borderwidth: 5,
      borderColor: '#123456',
      addFill: false,
      fillcolor: '#abcdef',
    };
    const rectangleTool = new RectangleTool(null);
    const state = rectangleTool.deserializeState(serializedObject);

    expect(state).toMatchObject(serializedObject);
  });
});
