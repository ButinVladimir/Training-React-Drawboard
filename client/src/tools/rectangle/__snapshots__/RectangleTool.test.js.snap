// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RectangleTool returns options element 1`] = `
<BaseFigureToolOptions
  addBorder={true}
  addFill={true}
  borderColor="#000000"
  borderWidth={1}
  fillColor="#ff0000"
  onChangeAddBorder={[Function]}
  onChangeAddFill={[Function]}
  onSelectBorderColor={[Function]}
  onSelectBorderWidth={[Function]}
  onSelectFillColor={[Function]}
  widths={
    Array [
      1,
      2,
      3,
      4,
      5,
      10,
      20,
      50,
    ]
  }
/>
`;
